facades/TeraEd/src/main/java/org/terasology/editor/subsystem/LwjglPortlet.java

```diff
@@ -0,0 +1,265 @@
+            }
+
+            @Override
+            public void paintGL() {
+                if (!((TerasologyEngine) engine).tick()) {
+                    // ignore?
+                }
+            }
+        };
+    }
+
+    public AWTGLCanvas getCanvas() {
+        return this.canvas;
+    }
+
+    private void initBuffer() {
+        logger.info("Initializing display (if last line in log then likely the game crashed from an issue with your " +
+                "video card)");
+
+        if (!config.isVSync()) {
+            GLFW.glfwSwapInterval(0);
+        }
+
+        try {
+            String root = "org/terasology/icons/";
+            ClassLoader classLoader = getClass().getClassLoader();
+
+            BufferedImage icon16 = ImageIO.read(classLoader.getResourceAsStream(root + "gooey_sweet_16.png"));
+            BufferedImage icon32 = ImageIO.read(classLoader.getResourceAsStream(root + "gooey_sweet_32.png"));
+            BufferedImage icon64 = ImageIO.read(classLoader.getResourceAsStream(root + "gooey_sweet_64.png"));
+            BufferedImage icon128 = ImageIO.read(classLoader.getResourceAsStream(root + "gooey_sweet_128.png"));
+            GLFWImage.Buffer buffer = GLFWImage.create(4);
+            buffer.put(0, LwjglGraphicsUtil.convertToGLFWFormat(icon16));
+            buffer.put(1, LwjglGraphicsUtil.convertToGLFWFormat(icon32));
+            buffer.put(2, LwjglGraphicsUtil.convertToGLFWFormat(icon64));
+            buffer.put(3, LwjglGraphicsUtil.convertToGLFWFormat(icon128));
+
+        } catch (IOException | IllegalArgumentException e) {
+            logger.warn("Could not set icon", e);
+        }
+
+        display.setDisplayModeSetting(config.getDisplayModeSetting());
+    }
+
+    private void initOpenGL(Context currentContext) {
+        logger.info("Initializing OpenGL");
+        LwjglGraphicsUtil.checkOpenGL();
+        LwjglGraphicsUtil.initOpenGLParams();
+        if (config.getDebug().isEnabled()) {
+            try {
+                GL43.glDebugMessageCallback(new DebugCallback(), MemoryUtil.NULL);
+            } catch (IllegalStateException e) {
+                logger.warn("Unable to specify DebugCallback to receive debugging messages from the GL.");
+            }
+        }
+        currentContext.put(ShaderManager.class, new ShaderManagerLwjgl());
+    }
+
+    public void initInputs() {
+        final InputSystem inputSystem = context.get(InputSystem.class);
+        Preconditions.checkNotNull(inputSystem);
+        canvas.addKeyListener(new KeyListener() {
+            @Override
+            public void keyTyped(KeyEvent e) {
+                ((LwjglKeyboardDevice) inputSystem.getKeyboard()).glfwCharCallback(0, e.getKeyChar());
+            }
+
+            @Override
+            public void keyPressed(KeyEvent e) {
+                ((LwjglKeyboardDevice) inputSystem.getKeyboard()).glfwKeyCallback(0, e.getKeyChar(),
```

Comment:
Swing's key codes equal glfw's key codes?

Reply:
Partially. Characters equals directly, special keys will have to be converted. Didn't do this yet. Same applies to mouse events - idea here is, that there will be for example mode in editor, where you click on it and mouse will go to exclusive mode to test gameplay until ESC for example... Or something like that.